### PR TITLE
Backend: Added logic to renew auth token if expired

### PR DESCRIPTION
**Pull request by vignesh-reddy**
issue https://github.com/alexanderzobnin/grafana-zabbix/issues/799

Added logic to renew the auth token if expired and the default number of attempts allowed is set to 3.

### CODE REVIEW

File: pkg/zabbix_api.go [hunk 1]
```diff
@@ -112,16 +112,22 @@ func (ds *ZabbixDatasource) BuildResponse(result *simplejson.Json) (*datasource.
...
+
+	for attempt := 0; attempt <= 3; attempt++ {
+		// Authenticate
 		auth, err := ds.loginWithDs(ctx, dsInfo)
 		if err != nil {
 			return nil, err
 		}
 		zabbixAuth = auth
+		result, err = ds.zabbixAPIRequest(ctx, zabbixUrl, method, params, zabbixAuth)
+		if !isNotAuthorized(err.Error()) {
```
Comment (alexanderzobnin) [inline pkg/zabbix_api.go]: `err` is always `nil` at this point, so it causes a panic

File: pkg/zabbix_api.go [hunk 2]
```diff
@@ -112,16 +112,22 @@ func (ds *ZabbixDatasource) BuildResponse(result *simplejson.Json) (*datasource.
...
+	var result *simplejson.Json
+	var err error
+
+	for attempt := 0; attempt <= 3; attempt++ {
+		// Authenticate
 		auth, err := ds.loginWithDs(ctx, dsInfo)
 		if err != nil {
 			return nil, err
 		}
 		zabbixAuth = auth
```
Comment (alecxvs) [inline pkg/zabbix_api.go]: This appears to be stored at the package level. If I understand the backend API correctly, we may be handling requests from multiple different Zabbix datasources during this runtime though. @alexanderzobnin

Comment (alexanderzobnin) [reply pkg/zabbix_api.go]: @alecxvs yes, you're right. This should be stored in datasource-specific object. I think we should create a cache instance for each datasource.

Comment (alexanderzobnin) [reply pkg/zabbix_api.go]: I think we can leave it as is for now, but I'll create an issue for moving things into datasource-specific cache.

File: pkg/zabbix_api.go (hunk 1 above)
Comment (vignesh-reddy) [reply pkg/zabbix_api.go]: Updated 4f5f5ec

File: pkg/zabbix_api.go (hunk 2 above)
Comment (alexanderzobnin) [reply pkg/zabbix_api.go]: @alecxvs I created an issue #812

File: pkg/zabbix_api.go [hunk 3]
```diff
@@ -112,16 +112,22 @@ func (ds *ZabbixDatasource) BuildResponse(result *simplejson.Json) (*datasource.
...
 	zabbixUrl := dsInfo.GetUrl()
 
-	// Authenticate first
-	if zabbixAuth == "" {
+	var result *simplejson.Json
+	var err error
+
+	for attempt := 0; attempt <= 3; attempt++ {
+		// Authenticate
 		auth, err := ds.loginWithDs(ctx, dsInfo)
```
Comment (alexanderzobnin) [inline pkg/zabbix_api.go]: In this case, data source tries to login at every request. We should use stored `zabbixAuth` token and renew it only if "not authorized" error happens.

Comment (vignesh-reddy) [reply pkg/zabbix_api.go]: Good catch, Updated https://github.com/alexanderzobnin/grafana-zabbix/pull/808/commits/371c68dd2e20fe720f565785000145f09ec6f539

**Pull request merged**